### ISSUE TITLE
Cleanup Duplicated Code in fragment shaders

### ISSUE THREAD

**Issue by cancan101** [FNNDSC/ami#220]
There is a fair bit of duplicated code between the `vr` and the `data` renderers both in the class itself and in the generated code. Further in one `getIntensity` has been factored out as a function whereas in the other it has not been.

**Comment by NicolasRannou** [comment]
Yes - as you mentionned getIntensity is a good example of things that should be cleaned up and moved out in a helper all shaders can used. In general I do not like very much way shaders codes is implemented and open to suggestions! The main requirement is that we want to be able to generate "custom" shaders "on-demand". (I can detail later if you are interested)

**Comment by NicolasRannou** [comment]
FYI we use to use glslify to generate shaders, that worked great but didn't allow us to build "custom" shaders "on demand".

## Custom shaders
If we pass a 16 bits datasets to the shaders, the "unpack" function would by default unpack 16 bits.

Main reason is performance, to cut out as many as possible unnecessary checks.
```
# before:
unpack (packedValue) {
  if (dataType = 8) {
    unPacked = ...
  } else if (dataType == 16) {
    unpacked = ...
  } ...

  return unPacked;
}

# now:
unpack (packedValue) {
  unpacked = ...
  return unPacked;
}
```

## On demand

That is tightly related to the custom shaders: we do not know we type of data will be loaded, which kind of interpolation the user wants to apply, etc. before hand so we do need to generate the shader code on the client side, once the parameters have been defined.

**Comment by cancan101** [comment]
I'm not sure I follow what the limitations of glslify were. It looks like st the very least parameter exist in require statements https://github.com/glslify/glslify/blob/master/README.md#passing-references-between-modules

**Comment by NicolasRannou** [comment]
So the initial main limitation was that "glslify" was (is?) not able to generate the GLSL code on the fly, in the client, but had to compile the shaders before hand while compiling AMI. That didn't allow us to generate "custom shaders" on the fly in the client. If that is now possible that is a big win and we should definitely look back into it -

Makes sense?

**Comment by cancan101** [comment]
Yes that makes sense. That being said, what is the issue with that limitation? Ie how much customization of the shaders occurs?

**Comment by NicolasRannou** [comment]
So the only reason really was to reduce the number of "else/if" statements in the shaders code in attempt to make it run faster (especially the volume rendering with interpolation ON).

Performance improvement was never formally tested though. (it "seemed" faster)

On the other hand GLSL code is now much less readable so honestly the more I think about it the more it makes sense to switch back to GLSLIFY.

In your experience are the else/if statements expensive in the "raycasting VR" context?
